Yoast/wordpress-seo
# Move Restore Site functionality into its own tab and submit network form via AJAX

## Summary

This PR can be summarized in the following changelog entry:

* Move the network's Restore Site functionality into its own tab and submit the network settings form via AJAX.

## Relevant technical choices:

### Part 1: Moving the Restore Site functionality
* First, I took the existing Restore Site form (which was currently printed via a footer hook) and moved the markup into its own template (`views/tabs/network/restore-site.php`). I removed the wrapping form tags, since the entire page already is a form.
* When the form submits, the `action` parameter is responsible for what happens (update settings or restore site?). For the regular settings update submission, that `action` is part of a hidden input. However, for other special use-cases, such as restoring a site, I figured it would make sense to put the `action` on the respective submit button: That makes sense because that `action` value is only submitted if that very button is clicked to submit the form.
* At this point, the functionality was already working correctly using a regular submission.

### Part 2: Using AJAX for the form
* In order to not require duplicate logic, I outsourced two parts of `Yoast_Network_Admin` into more specific private methods: First, checking whether the current action request is valid and permitted (checking nonce and capabilities). Second, dealing with the action response (making notices available to the user, using the settings errors API).
    * Verifying the request: For a regular request, `check_admin_referer()` is called. If the user has insufficient capabilities, `wp_die()` is called. For an AJAX request, the same code calls `check_ajax_referer()` instead. If the user has insufficient capabilities, `wp_die()` is called as well, but with only a -1 to resemble how other AJAX functions handle that.
    * Providing the user with feedback: For a regular request, the current settings errors are persisted, and then the user is redirected back to where they came from. For an AJAX request, the same code instead returns the settings errors directly as JSON.
* With the above foundation for AJAX functionality, it is possible to use the same methods for both regular requests and AJAX requests.
* The AJAX hooks however cannot be registered in the `Yoast_Network_Admin` class. That is because WP AJAX fires in a different way where the plugin does not load that class automatically. I therefore introduced a new `WPSEO_WordPress_AJAX_Integration` with a `register_ajax_hooks()` method and have `Yoast_Network_Admin` implement it. The method is then called in `admin/ajax.php`. It would make sense to migrate further classes to use that interface in the future.
* With AJAX in place on the PHP end, the remaining part was the JS file. I introduced a new `wp-seo-network-admin.js` for that purpose. It catches every form submission, detects the current action and then re-routes the submission to happen via AJAX. From the JSON response given, the settings errors are then dynamically inserted into the page.

## Test instructions

This PR can be tested by following these steps:

* You need to set up a multisite and make yourself a network administrator. Then the plugin needs to be activated network-wide.
* There are two parts of testing, like indicated in the Technical Choices section above. First, please test the functionality without AJAX. You need to temporarily comment out the contents of the `Yoast_Network_Admin::enqueue_assets()` method for that.
* Now go to the Yoast Network Settings screen. Change the two settings around, submit the form and see whether they save and persist correctly.
* Go into the regular site admin now and change a few settings of the plugin so that they are different from their defaults. After that, return to the Yoast Network Settings screen. Go into the Restore Site tab, select the site where you changed the settings, and submit the form. Verify that the settings you had changed before on the site are now back to their defaults.
* Now uncomment what you previously commented out so that you can test the AJAX behavior.
* Do exactly the same as before, in order to test the network settings form submission and the restore site form submission. In addition to what was highlighted above, pay extra attention to whether the feedback from those actions shows up correctly.
* The overall experience with AJAX should be smoother, but otherwise (except for the lack of a new pageload) similar to what happens without the AJAX functionality.

## Quality assurance


Fixes #9689
